File: compiler/bench/interpolsearch.c
// expected: 499500000
// Interpolation search benchmark - search 1M times in sorted array

#include <stdio.h>

#define SIZE 1000
#define ITERS 1000000

int arr[SIZE];

int interp_search(int val) {
    int lo = 0, hi = SIZE - 1;
    while (lo <= hi && val >= arr[lo] && val <= arr[hi]) {
        if (arr[hi] == arr[lo]) {
            if (arr[lo] == val) return lo;
            return -1;
        }
        // pos = lo + ((val - arr[lo]) * (hi - lo)) / (arr[hi] - arr[lo])
        int pos = lo + ((long)(val - arr[lo]) * (hi - lo)) / (arr[hi] - arr[lo]);
        if (arr[pos] == val) return pos;
        if (arr[pos] < val) {
            hi = pos - 1;
        } else {
            lo = pos + 1;
        }
    }
    return -1;
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = i;
    }
}

long bench(void) {
    long sum = 0;
    for (int i = 0; i < ITERS; i++) {
        sum += interp_search(i % SIZE);
    }
    return sum;
}

int main(void) {
    init_arr();
    printf("%ld\n", bench());
    return 0;
}

File: compiler/bench/interpolsearch.fs
\ expected: 499500000
\ Interpolation search benchmark - search 1M times in sorted array

1000 constant SIZE
1000000 constant ITERS
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: interp-search ( val -- idx|-1 )
  0 SIZE 1-             \ val lo hi
  begin
    2dup <= 3 pick 2 pick arr@ >= and 3 pick over arr@ <= and
  while
    \ pos = lo + ((val - arr[lo]) * (hi - lo)) / (arr[hi] - arr[lo])
    over arr@ 3 pick arr@ - dup 0= if
      \ arr[lo] == arr[hi]
      drop 2dup = if nip nip exit then
      2drop drop -1 exit
    then
    >r                    \ val lo hi ; R: arr[hi]-arr[lo]
    3 pick 2 pick arr@ - \ val lo hi (val-arr[lo])
    over 2 pick - *       \ val lo hi (val-arr[lo])*(hi-lo)
    r> /                  \ val lo hi pos-offset
    2 pick +              \ val lo hi pos
    dup arr@ 4 pick = if
      nip nip nip exit
    then
    dup arr@ 4 pick < if
      \ arr[pos] < val
      nip 1-              \ val lo pos-1
    else
      \ arr[pos] > val
      rot drop 1+ swap    \ val pos+1 hi
    then
  repeat
  2drop drop -1 ;

: init-arr ( -- )
  SIZE 0 do i i arr! loop ;

: bench ( -- sum )
  0
  ITERS 0 do
    i SIZE mod interp-search +
  loop ;

: main
  init-arr
  bench . cr ;
